client/slack/invite_form.js

```diff
@@ -1,27 +1,100 @@
+                                onChange={ e => this.setState({ email: e.target.value }) }
+                                value={ email } required />
+                    <Form.Checkbox label={ <label>I agree to the <a href="/terms/">Terms and Code of Conduct</a></label> }
+                                   onChange={ () => this.setState({ acceptTos: !acceptTos }) }
+                                   checked={ acceptTos } required />
+                    <Message error header='Could you check something?' list={ this.formatErrors(errors) }
+                             content={
+                                 <small>
+                                     If this seems to be an error, please email us at&nbsp;
+                                     <a href="mailto:pythondev.slack@gmail.com">pythondev.slack@gmail.com</a>
+                                 </small>} />
+                    <Form.Button onClick={ this.onSubmit } disabled={ this.shouldButtonDisable() } primary>
+                        Submit
+                    </Form.Button >
+                </Form>
+            </div>
+        );
+    }
+
+    onSubmit = async (e) => {
+        e.preventDefault();
+
+        const { email, acceptTos } = this.state;
+
+        this.setState({ errors: {}, loading: true, success: false });
+        try {
+            const inviteTask = await $.ajax({
+                url: '/slack/',
+                method: 'POST',
+                data: {
+                    csrfmiddlewaretoken: document.getElementsByName('csrfmiddlewaretoken')[0].value,
+                    email: email,
+                    accept_tos: acceptTos,
                 },
-                accept_tos: {
-                    identifier: 'accept_tos',
-                    rules: [
-                        {
-                            type: 'checked',
-                            prompt: 'You must agree to the terms'
-                        }
-                    ]
-                }
+            });
+
+            const taskResult = await retrieveTaskResult(inviteTask.task_id);
+            const newState = (taskResult.error !== undefined)
+                ? { errors: { email: [{message: taskResult.error}] } }
+                : { email: '', acceptTos: false, success: true };
+            this.setState(newState);
+        } catch (e) {
+            if ([400, 429].indexOf(e.status) !== -1) {
+                this.setState({ errors: e.responseJSON });
+            } else {
+                console.error('Unknown error:', e);
+                alert('Sorry, something went wrong. Please refresh and try again.');
             }
-        });
-});
+        }
+        this.setState({ loading: false });
+    };
+
+    shouldButtonDisable = () => {
+        const { email, acceptTos } = this.state;
+        return email === '' || !acceptTos;
+    };
+
+    formHasError = () => {
+        return Object.keys(this.state.errors).length > 0 && this.state.errors.constructor === Object;
```

Comment:
Instead of looping through the keys, and because you already have jQuery in scope, you could just use the `isEmptyObject` method

Reply:
My initial intent was to strip jquery, but since I didn't follow throw this is cleaner. Later I may pull in lodash if the jquery status changes